### DISCUSSION TITLE
Where is get(key, default)?

### DISCUSSION THREAD

**Discussion by gatopeich** [nlohmann/json#2278] [Ideas]
#### Improvement

The library would be more usable if it offered `get(key, default=null)` interface.

#### Why

A great feature of this library is bringing the power of Python/Javascript objects to C++.

Only it lacks `object.get(key, default=null)` interface, which is such an useful feature!

Turns out that the functionality is actually there, but the choice of name "value" is not adequate. It takes a bit of research to find, actually I thought it was not there for a while and had to implement it myself.

It also happens that the semantics of "get" and "value" are swapped: each one means what the other says!

Since swapping "get()" with "value()" would break existing code, I hereby propose to make them work both ways:

* `get(key, default) == value(key, default)`
* `value() == get()`

#### Please describe the steps to reproduce the issue.

1. Start using this great library
2. Where is `object.get(key, default)` ???
3. Can't find it, end up writing `object.contains(key) ? object[key] : default`

**Comment by nlohmann** [comment]
I don't think changing the names is a good idea now, as it would be a breaking change and probably add more confusion than resolving it.

However, this is the rationale for the names:

- `T basic_json::get<T>()` was motivated by the `std::get` function from `std::variant` or `std::tuple`. You provide a type and you get the value.
- That name gone, we had to find something else for "give me an object value with a default value". Hence `value`.

Sorry for the inconvenience.

**Reply by gatopeich** [reply]
Thanks for the rationale.

Still I see no conflict with having an overload `T basic_json::get<T>(KeyT key, T default=null) const`.

Is it as hard as an `ordered_json`? :slightly_smiling_face:

**Reply by nlohmann** [reply]
Having two functions doing the same thing would be even more confusing.

**Reply by gatopeich** [reply]
Maybe, for maintainers.

The confusion for _users_ arises from not finding "get(key)" in the first place.

If the reason to have "value" was that "get" was gone, there is no reason.

**Comment by cxwx** [comment]
still no get（key） function for default value？

**Reply by nlohmann** [reply]
It's called value.

**Reply by cxwx** [reply]
thanks

**Reply by cxwx** [reply]
@nlohmann 
if I have an array
```
{"source":
[
{"a":“string”},
{"a":null},
...
]
}
```
I'd love to use a for cycle
```
for (auto i : json["source"]) {
if(json.at("a").is_null()){ // how can I remove this line?
...
}
else
{
int k = json.value("a", “-1”);
}
}
```
Can I set the null as default value?
